src/org/exist/storage/XQueryPool.java

```diff
@@ -160,78 +155,57 @@ private synchronized void returnObject(final Source source, final CompiledXQuery
         }
     }
 
-    private Object borrowObject(final DBBroker broker, final Source source) {
-        final Source key;
-        final CompiledXQuery query;
-        synchronized (this) {
-            final int idx = getIndex(source);
-            if (idx < 0) {
-                return null;
-            }
-            key = (Source) keys[idx];
-            int validity = key.isValid(broker);
-            if (validity == Source.UNKNOWN) {
-                validity = key.isValid(source);
-            }
+    private Object borrowObject(final Source source) {
```

Comment:
@shabanovd Great looks good. Can you squash your fixup commits please into the base commit?